nova/scheduler/filters/shard_filter.py

```diff
@@ -114,11 +125,106 @@ def _get_shards(self, project_id):
 
         return self._PROJECT_SHARD_CACHE.get(project_id)
 
-    def host_passes(self, host_state, spec_obj):
+    def _get_k8s_shard(self, spec_obj):
+        """Returns the dominant shard of a K8S cluster.
+
+        Returns None if the request is not for an instance that's part of
+        a K8S cluster, or if this is the first instance of a new cluster.
+        """
+        if (spec_obj.flavor.name.startswith(HANA_PREFIX) or
+                utils.request_is_resize(spec_obj)):
+            return None
+        elevated = nova_context.get_admin_context()
+        build_request = None
+        instance = None
+
+        def _get_tags():
+            return build_request.tags if build_request \
+                else instance.tags
+
+        def _get_metadata():
+            return build_request.instance.metadata if build_request \
+                else instance.metadata
+
+        check_type = spec_obj.get_scheduler_hint('_nova_check_type')
```

Comment:
Is this safe i.e. do we have a build request every time we do not set `_nova_check_type`? I see `resize`, `live-migrate` and `rebuild` setting it, but e.g. not `unshelve`.

Would it make sense to try and get the `BuildRequest` if we don't have a `check_type`, but fall back to getting the instance instead?